### PR TITLE
audit: error on archived repos

### PR DESCRIPTION
**Pull request by SMillerDev**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
Since we require all software to be maintained we should avoid archived repos. This will warn about that.

### CODE REVIEW

**Comment by Bo98** [comment]
It might not matter much since it’s `--strict`, but any thoughts on making it not flag formulae marked as deprecated or disabled?

**Comment by SMillerDev** [comment]
Seems like a good idea.

File: Library/Homebrew/utils/notability.rb [hunk 1]
```diff
@@ -5,12 +5,14 @@
 module SharedAudits
   module_function
 
+  def github_repo_data(user, repo)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/notability.rb:8]: Wondering if this should be cached so we're not querying the same repo again and again. Thoughts?

File: Library/Homebrew/dev-cmd/audit.rb [hunk 2]
```diff
@@ -522,6 +522,26 @@ def audit_bottle_disabled
       problem "Formulae in homebrew/core should not use `bottle :disabled`"
     end
 
+    def audit_github_archived
+      user, repo = get_repo_data(%r{https?://github\.com/([^/]+)/([^/]+)/?.*}) if @strict
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb:526]: Probably needs a `return unless @online`?

**Comment by SMillerDev** [comment]
Disabled shouldn't be tested at all, but I added a return if deprecated. Otherwise CI would fail if you deprecate an archived repo.

**Comment by Bo98** [comment]
> Disabled shouldn't be tested at all

Well this audit wouldn't run when making _any_ modification to an existing formula as we only do `--strict` checks on new formulae.

**Comment by SMillerDev** [comment]
Why not online?

**Comment by Bo98** [comment]
`--online` is run, yes. But there's `if @strict` here too which I'm talking about.

**Comment by SMillerDev** [comment]
Ah, I get the problem now. I'm removing the strict requirement.

File: Library/Homebrew/utils/notability.rb [hunk 3]
```diff
@@ -5,12 +5,18 @@
...
+    return @github_data unless @github_data.nil?
+
+    @github_data = GitHub.repository(user, repo)
+
+    @github_data
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/notability.rb:13]:
```suggestion
    @github_repo_data ||= {}
    @github_repo_data["#{user}/#{repo}"] ||= GitHub.repository(user, repo)
```

File: Library/Homebrew/utils/notability.rb [hunk 4]
```diff
@@ -25,11 +31,19 @@ def github(user, repo)
     "GitHub repository too new (<30 days old)"
   end
 
-  def gitlab(user, repo)
+  def gitlab_repo_data(user, repo)
+    return @gitlab_data unless @gitlab_data.nil?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/notability.rb:35]: Need to do something similar to `@github_repo_data` above or it'll reuse the same data for every repo.

File: Library/Homebrew/utils/notability.rb [hunk 5]
```diff
@@ -5,12 +5,29 @@
...
+  def gitlab_repo_data(user, repo)
+    @gitlab_repo_data ||= {}
+    return @gitlab_repo_data["#{user}/#{repo}"] unless @gitlab_repo_data["#{user}/#{repo}"].nil?
+
+    out, _, status= curl_output("--request", "GET", "https://gitlab.com/api/v4/projects/#{user}%2F#{repo}")
+    return unless status.success?
+
+    @gitlab_repo_data["#{user}/#{repo}"] = JSON.parse(out)
+
+    @gitlab_data["#{user}/#{repo}"]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/notability.rb:26]:
```suggestion
    @gitlab_repo_data ||= {}
    @gitlab_repo_data["#{user}/#{repo}"] ||= begin 
      out, _, status= curl_output("--request", "GET", "https://gitlab.com/api/v4/projects/#{user}%2F#{repo}")
      return unless status.success?

      JSON.parse(out)
    end

    @gitlab_data["#{user}/#{repo}"]
```

Comment (SMillerDev) [reply Library/Homebrew/utils/notability.rb:26]: Ruby is magic.

File: Library/Homebrew/utils/notability.rb [hunk 6]
```diff
@@ -5,13 +5,30 @@
...
-  def github(user, repo)
-    begin
-      metadata = GitHub.repository(user, repo)
-    rescue GitHub::HTTPNotFoundError
-      return
+  def github_repo_data(user, repo)
+    @github_repo_data ||= {}
+    @github_repo_data["#{user}/#{repo}"] ||= GitHub.repository(user, repo)
+
+    @github_data["#{user}/#{repo}"]
```
Comment (Bo98) [inline Library/Homebrew/utils/notability.rb:12]:
```suggestion
    @github_repo_data["#{user}/#{repo}"]
```

File: Library/Homebrew/utils/notability.rb [hunk 7]
```diff
@@ -5,13 +5,30 @@
...
+  def gitlab_repo_data(user, repo)
+    @gitlab_repo_data ||= {}
+    @gitlab_repo_data["#{user}/#{repo}"] ||= begin
+      out, _, status= curl_output("--request", "GET", "https://gitlab.com/api/v4/projects/#{user}%2F#{repo}")
+      return unless status.success?
+
+      JSON.parse(out)
     end
 
+    @gitlab_data["#{user}/#{repo}"]
```
Comment (Bo98) [inline Library/Homebrew/utils/notability.rb:26]:
```suggestion
    @gitlab_repo_data["#{user}/#{repo}"]
```

**Comment by dawidd6** [comment]
```
==> brew audit kafkacat --online --git --skip-style
==> FAILED
Error: undefined method `[]' for nil:NilClass
/usr/local/Homebrew/Library/Homebrew/utils/notability.rb:12:in `github_repo_data'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:531:in `audit_github_repository_archived'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:893:in `block in audit'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:886:in `each'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:886:in `audit'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:132:in `block in audit'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:117:in `each'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:117:in `audit'
/usr/local/Homebrew/Library/Homebrew/brew.rb:111:in `<main>'
```

**Pull request merged**